### DISCUSSION TITLE
Computing fallback locales

### DISCUSSION THREAD

**Discussion by samhastings** [symfony/symfony#51748] [Q&A]
Based on my research I'm going to assume the answer to this question is "no", however I figured there was no harm in asking just in case I've missed something!

According to the [Fallback Translation Locales section of the translation documentation](https://symfony.com/doc/current/translation.html#fallback-translation-locales), Symfony will fall back from `es_AR` to `es_419`, then to `es`, then to any fallbacks or the default locale defined in `config/packages/translation.yaml`.

I kind of need to implement this logic in my own application to determine a chain of fallback locales in order to query the correct version of a stored Doctrine entity. So given an input of `es_AR`, it would be really useful to be able to get the parent/grandparent/etc locales shown above so I can query each of these in turn until one returns an entity.

The `Symfony\Component\Intl\Locale` class provides a `getParent()` method but this simply removes the suffix and jumps directly from `es_AR` to `es`, skipping the `es_419` locale.

The `Symfony\Component\Translation\Translator` class includes a `computeFallbackLocales()` method but this is protected and therefore not part of the translation component’s API.

Is there a way I can achieve this using existing Symfony APIs, or am I going to have to implement this myself?

**Comment by MatTheCat** [comment]
"no" :grin: 

This is a translator’s implementation detail so it makes sense it isn’t exposed. Also the fact you “kind of need” the logic maybe implies this wouldn’t work for you.

> The `Symfony\Component\Intl\Locale` class provides a `getParent()` method but this simply removes the suffix and jumps directly from `es_AR` to `es`, skipping the `es_419` locale.

I guess you meant `getFallback` instead of `getParent`? These are different things:

- https://unicode-org.github.io/icu/userguide/locale/#fallback
- https://www.unicode.org/reports/tr35/tr35.html#Parent_Locales

You could open an issue to add parent locales to Symfony’s Intl resource bundles, but in the meantime you have two options:

- use the `parents.json` file like the translator does
- use intl’s `ResourceBundle`: `(new \ResourceBundle('es_AR', null, false))->get('%%Parent', false)`
